### PR TITLE
File delete fix on high load

### PR DESCRIPTION
**Pull request by sartor**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌

On high load projects sometimes there is an error trying to delete file. This happens because of not atomic check if file exists and actual delete:
```php
if (!is_file($file)) {
    return true;
}

return @unlink($file);
```

In old yii2 file cache result of key delete not throw any error.
This PR tries to return false only if `unlink` call error caused by any reason except file already deleted
I can't do simple test to prove fix, but I've checked it by calling first `unlink` just after `is_file()` check

### CODE REVIEW

File: src/FileCache.php [hunk 1]
```diff
@@ -168,7 +168,17 @@ public function delete(string $key): bool
             return true;
         }
 
-        return @unlink($file);
+        $result = @unlink($file);
+        // Check if error was because of file was already deleted by another process on high load
+        if ($result === false) {
+            $lastError = error_get_last();
+            if (str_ends_with($lastError['message'] ?? '', 'No such file or directory')) {
```
Comment (vjik) [inline src/FileCache.php:175]:
```suggestion
            if (!is_file($file)) {
```

Maybe so more simplify?

Comment (sartor) [reply src/FileCache.php:175]: No, after unlink, this file can be already created by another process again

**Comment by rustamwin** [comment]
Need tests.

**Comment by sartor** [comment]
> Need tests.

Any ideas how it can be done?

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
It's hard to test race conditions. I think we can skip these for the case. @sartor would you please add a line for CHANGELOG

**Review by vjik — CHANGES_REQUESTED** [review]
@sartor Can you add line to changelog?

**Review by vjik — APPROVED** [review]

**Comment by vjik** [comment]
Seems, similar problem may occur in method `removeCacheFiles()`:

https://github.com/yiisoft/cache-file/blob/6c0eda63e048ecb20808bfd3d2c9a836ce98b670/src/FileCache.php#L391

https://github.com/yiisoft/cache-file/blob/6c0eda63e048ecb20808bfd3d2c9a836ce98b670/src/FileCache.php#L395

**Comment by samdark** [comment]
Yes, likely.

**Comment by sartor** [comment]
Yes. I'm using on prod my fork for a week, and fixed line weren't produced any new errors.
But there are similar errors left in other places. I'll fix them later in another PR

**Pull request merged**